In [2]:
import pandas as pd
import numpy as np

#Reading the dataset
#Patents of class 424 by year
df = pd.read_csv('patent2005.csv', low_memory=False)

#Dropping collumns not necessary for processing
patent2005 = df.drop(['appcit_app_number','appcit_category','appcit_date','appcit_kind',
                      'appcit_sequence','assignee_sequence','uspc_sequence','wipo_sequence',
                     'patent_kind','patent_abstract','patent_type','app_date','inventor_lastknown_country','assignee_location_id'], axis=1)


#Changing position of columns for a better understanding
patent2005 = patent2005.set_index(['patent_number','uspc_mainclass_id','uspc_mainclass_title',
                                   'patent_date','patent_title','patent_year'])
patent2005 = patent2005.reset_index()

#Displaying 
patent2005.head(9)

patent_number uspc_mainclass_id  \
0       6837925               106   
1       6837925               106   
2       6837925               106   
3       6837925               106   
4       6837925               106   
5       6837925               424   
6       6837925               106   
7       6837925               106   
8       6838054               204   

                                uspc_mainclass_title patent_date  \
0                  Compositions:  coating or plastic  2005-01-04   
1                  Compositions:  coating or plastic  2005-01-04   
2                  Compositions:  coating or plastic  2005-01-04   
3                  Compositions:  coating or plastic  2005-01-04   
4                  Compositions:  coating or plastic  2005-01-04   
5  Drug, bio-affecting and body treating composit...  2005-01-04   
6                  Compositions:  coating or plastic  2005-01-04   
7                  Compositions:  coating or plastic  2005-01-04   
8             Chemistry:  electrical and wave energy  2005-01-04   

                                        patent_title  patent_year  app_number  \
0  Coloring pearlescent flake pigment, manufactur...         2005    10447276   
1  Coloring pearlescent flake pigment, manufactur...         2005    10447276   
2  Coloring pearlescent flake pigment, manufactur...         2005    10447276   
3  Coloring pearlescent flake pigment, manufactur...         2005    10447276   
4  Coloring pearlescent flake pigment, manufactur...         2005    10447276   
5  Coloring pearlescent flake pigment, manufactur...         2005    10447276   
6  Coloring pearlescent flake pigment, manufactur...         2005    10447276   
7  Coloring pearlescent flake pigment, manufactur...         2005    10447276   
8  Biochemical devices and their methods of manuf...         2005     9673538   

   app_type assignee_country assignee_first_name  ... inventor_country  \
0        10               JP                None  ...               JP   
1        10               JP             Yasushi  ...               JP   
2        10               JP                None  ...               JP   
3        10               JP                None  ...               JP   
4        10               JP                None  ...               JP   
5        10               JP                None  ...               JP   
6        10               JP                None  ...               JP   
7        10               JP                None  ...               JP   
8         9               GB                None  ...               GB   

  inventor_first_name inventor_id inventor_last_name inventor_sequence  \
0             Yasushi   4340697-2               Kubo                 0   
1              Hideji   5244731-5             Kagawa                 1   
2             Yasushi   4340697-2               Kubo                 0   
3             Yasushi   4340697-2               Kubo                 0   
4             Yasushi   4340697-2               Kubo                 0   
5             Yasushi   4340697-2               Kubo                 0   
6             Yasushi   4340697-2               Kubo                 0   
7             Yasushi   4340697-2               Kubo                 0   
8        James Robert   6838054-1            Durrant                 0   

  uspc_subclass_id                                uspc_subclass_title  \
0          106/486      Clay or material derived therefrom containing   
1          106/415         Mica, shell, scale, platelet, or lamellate   
2          106/416               Clay containing or derived therefrom   
3          106/468  Clay or material derived from clay containing,...   
4          106/487                        Organic material containing   
5           424/69  Face or body powders for grooming, adorning, o...   
6          106/486      Clay or material derived therefrom containing   
7          106/486      Clay or material derived therefrom containing   


Patents are identified based on their `patent_number` which is unique for each patent. In the dataset below is all the information corresponding to the patent of number `6837925`.

We create an array containing only unique values of `patent_number`from the dataset `patent20005`. Then, based on all the unique and different values of this array we create a temporary dataset only displaying the information of a desired patent by specifying the patent-number. See line 6 of code below to understand how an arbitrary patent number is selected.

In [42]:
#Displaying only information on singular patent, selected by patent number
patent_number_list = patent2005['patent_number']
patent_number = patent_number_list.unique()
patent_title_list = patent2005['patent_title']
patent_title=patent_title_list.unique()
temp_df= patent2005.loc[(patent2005.patent_number==patent_number[3])]
temp_df['cited_patent_title']

53                 Catalytic hydroformylation of olefins
54     Technetium (Tc-99m)-arene complexes useful in ...
55                          Terpyridine chelating agents
56     Ether isonitriles and radiolabeled complexes t...
57     Process for preparing Tc-99m tris(isonitrile) ...
58     Process for preparing mono-n-alkylated polyaza...
59     Heterocyclic derivatives of DTPA used for magn...
60     Polychelating agents for image and spectral en...
61                                 Polychelant compounds
62     Complexes with .sup.99m Tc suitable for radiol...
63                       Radiolabelled peptide compounds
64     Amplifier molecules derived from diethylene tr...
65     Functionalized aza-crytand ligands for diagnos...
66     Calcium/oxyanion-containing particles with a p...
67     Low density microspheres and suspensions and t...
68     Low density microspheres and their use as cont...
69     TC or RE radionuclide labelled chelate, hexape...
70     Dendritic amplifier mole

Now we would like to know how many unique patents exist in this dataset to define a method capable of processing the entire dataset in order to create all the indicators.

In [14]:
#Total of patent in this dataset
len(patent_number)


2701

We can also identify certain features by processing the meta-data of a patent after creating the temporary dataset. For example, to how many main classes and subclasses the patent belongs and how many inventors worked on this project.

Now we would like to calculate de median age of all the cited patents, but some of the patents has void values in the column `cited_patent_date` as `None`, we need to make sure thos empy cells dont affect our method for generating the median age. By creating a mask, we filter the column to only process the cells that contain a readable date.

In [45]:

mask = (temp_df['cited_patent_date'] != 'None')
temp_df_valid = temp_df[mask]

if(temp_df_valid.size !=0):
    TCT = np.mean(pd.DatetimeIndex(temp_df_valid['cited_patent_date']).year).astype(int)
else:
    TCT = 0
TCT

1997

In [13]:


#Median Age of cited patents
TCT = np.mean(pd.DatetimeIndex(temp_df['cited_patent_date']).year).astype(int)

1993

In [317]:
#How many classes the patent belongs
total_mainclass = len(temp_df['uspc_mainclass_id'].unique())
print("Total de USPC \t",total_mainclass)

#How many inventors has a patent
total_inventors = len(temp_df['inventor_id'].unique())

print("Total de inventores \t",total_inventors)


#How many subclasses it belongs
total_subclasses = len(temp_df['uspc_subclass_id'].unique())
print("Total de subclases \t",total_subclasses)


Total de USPC 	 2
Total de inventores 	 2
Total de subclases 	 6


In [291]:
temp= patent2005.loc[(patent2005.patent_number==patent_number[0])]

countries=len(temp['assignee_country'].unique())
countries

1

The numerical value of the mainclass of the patent following the USPC clasification is used as another feature to characterize the patent.

In [266]:
temp= patent2005.loc[(patent2005.patent_number==patent_number[2])]

main_field=temp['uspc_mainclass_id'].value_counts()
mf=int(main_field.idxmax())

424

Generating the output indicators
As all the literature has shown, the obvious value to describe the emerginess of a technology is base on the number of future patents that will reference. We obtain this number by processing the `citedby_patent_date` and counting in 3 differentes periods of time how many patents have cited it.

In [194]:
temp= patent2005.loc[(patent2005.patent_number==patent_number[2])]

st_date= np.array(pd.to_datetime(temp['patent_date'].unique()).year +3)
mt_date= st_date+2
lt_date= mt_date+5
citedby = np.array(pd.to_datetime(temp['citedby_patent_date'].unique()).year)
FC3=sum(citedby<=st_date)
FC5=sum(citedby<=mt_date)
FC10=sum(citedby<=lt_date)

citedby

array([2006], dtype=int64)

Studies have found the positive relationship between coassignee and the value of a patent.This feature is 1 if a patent has more than one assignee, otherwise 0.

In [220]:
temp= patent2005.loc[(patent2005.patent_number==patent_number[1])]
assignee_total = len(temp['assignee_organization'].unique()) 

col=int(assignee_total>1)


0

In [292]:
def data_processing_per_year(patent2005):
    #Grouping by patent
    patent_number_list = patent2005['patent_number']
    patent_number = patent_number_list.unique()

    #Creating the returning DF
    ultimate_df = pd.DataFrame()

    
    for pnumber in patent_number:
        temp= patent2005.loc[(patent2005.patent_number==pnumber)]
        
        
        #Numerical value of the main field of the patent
        main_field=temp['uspc_mainclass_id'].value_counts()
        MF=int(main_field.idxmax())
        
        
        #Number of classes to which the patent belongs
        TS = len(temp['uspc_mainclass_id'].unique())
        
        #Number of subclasses to which the patent belongs
        total_subclasses = len(temp['uspc_subclass_id'].unique())
        
        #Number of inventors
        total_inventors = len(temp['inventor_id'].unique())
        
        
        #Calculating the mean age of cited patents
        mask = (temp['cited_patent_date'] != 'None')
        temp_df_valid = temp[mask]

        if(temp_df_valid.size !=0):
            TCT = np.mean(pd.DatetimeIndex(temp_df_valid['cited_patent_date']).year).astype(int)
        else:
            TCT = 0
        
        
        
        #Number of countries in which the patent has been registered
        countries=len(temp['assignee_country'].unique())
        
        
        #Binary value
        #describing if a patent was created in 
        #collaboration with multiple assignees
        assignee_total = len(temp['assignee_organization'].unique()) 
        col=int(assignee_total>1) #1 if true, 0 if false
        
        
        
        #Number of forward citations over the next 3, 5 and 10 years
        mask_citedby = (temp['citedby_patent_date'] != 'None')
        temp_citedby = temp[mask_citedby]
        citedby = np.array(pd.to_datetime(temp_citedby['citedby_patent_date'].unique()).year)
        st_date= np.array(pd.to_datetime(temp_citedby['patent_date'].unique()).year +3)
        mt_date= st_date+2
        lt_date= mt_date+5
        
        FC3=sum(citedby<=st_date)
        FC5=sum(citedby<=mt_date)
        FC10=sum(citedby<=lt_date)
      

        ultimate_df = ultimate_df.append({'patent_number' : pnumber, 
                                          'TS' : TS,
                                          'MF':MF,
                                          'INV':total_inventors,
                                          'SC':total_subclasses,
                                          'CS':countries,
                                          'COL':col,
                                          'TCT':TCT,
                                          'FC3':FC3,
                                          'FC5':FC5,
                                          'FC10':FC10
                                             },ignore_index=True)

    return ultimate_df


In [293]:
dff_2005 = data_processing_per_year(patent2005)

In [297]:
#dff_2005['CS'].value_counts()
dff_2005.set_index('patent_number')

COL   CS  FC10  FC3  FC5   INV     MF    SC     TCT   TS
patent_number                                                          
6837925        1.0  1.0   4.0  0.0  1.0   2.0  106.0   6.0     0.0  2.0
6838054        0.0  1.0   1.0  1.0  1.0   1.0  422.0  11.0  1995.0  6.0
6838073        1.0  1.0  28.0  3.0  7.0   2.0  424.0   9.0  1993.0  3.0
6838074        0.0  1.0   4.0  2.0  2.0   1.0  424.0   6.0  1997.0  3.0
6838075        0.0  1.0   7.0  0.0  0.0   9.0  128.0   7.0  1997.0  4.0
...            ...  ...   ...  ...  ...   ...    ...   ...     ...  ...
RE38827        0.0  1.0   6.0  1.0  3.0   4.0  424.0  17.0  1993.0  4.0
RE38912        0.0  1.0  24.0  1.0  3.0   2.0  424.0   5.0  1994.0  1.0
RE38913        0.0  1.0   7.0  1.0  4.0   1.0  424.0   6.0  1993.0  2.0
RE38916        1.0  3.0   1.0  1.0  1.0  12.0  424.0  10.0  1992.0  3.0
RE38919        0.0  1.0   9.0  0.0  4.0   2.0  424.0   3.0  1988.0  1.0

[2701 rows x 10 columns]